### PR TITLE
libdatachannel: init at 0.18.5

### PR DESCRIPTION
**Pull request by erdnaxe**
###### Description of changes

Motivation: Next major version of OBS Studio will depend on libdatachannel for WHIP (WebRTC) support.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - Built OBS Studio master with it, and successfully streamed using WHIP.
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:16]:
```suggestion
  version = "0.18.5";
```
New version has been released in the meantime

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:52]:
```suggestion
  postFixup = ''
    # Fix shared library path that will be incorrect on move to "dev" output
    substituteInPlace "$dev/lib/cmake/LibDataChannel/LibDataChannelTargets-release.cmake" \
      --replace "\''${_IMPORT_PREFIX}/lib" "$out/lib"
  '';
```
This pattern turns up a lot when running `git grep _IMPORT_PREFIX`.

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:23]:
```suggestion
```
Let's not vendor dependencies if possible, see postPatch suggestion below

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:39]:
```suggestion
```
See -DUSE_SYSTEM_USRSCTP suggestion

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:48]:
```suggestion
  postPatch = ''
    # TODO: Remove when updating to 0.19.x, and add
    # -DUSE_SYSTEM_USRSCTP=ON and -DUSE_SYSTEM_PLOG=ON to cmakeFlags instead
    mkdir -p deps/{usrsctp,plog}
    cp -r --no-preserve=mode ${srcOnly usrsctp}/. deps/usrsctp
    cp -r --no-preserve=mode ${srcOnly plog}/. deps/plog
  '';
```
That allows us to use the nixpkgs versions of usrsctp and plog. The advantage is that if we update usrsctp in nixpkgs, then libdatachannel will also get that updated version.

<details>
<summary>plog derivation scaffold (you can add me to maintainers if desired)</summary>

```
{ lib
, stdenv
, fetchFromGitHub
, cmake
}:

stdenv.mkDerivation rec {
  pname = "plog";
  version = "1.1.9";

  outputs = [ "out" "dev" ];

  src = fetchFromGitHub {
    owner = "SergiusTheBest";
    repo = pname;
    rev = version;
    hash = "sha256-CARTr1EEqXNJtp6XwHhr7aiRBiYR0cClzexpNCMaQJc=";
  };

  nativeBuildInputs = [
    cmake
  ];

  cmakeFlags = [
    "-DPLOG_BUILD_SAMPLES=NO"
  ];

  meta = with lib; {
    description = "Portable, simple and extensible C++ logging library";
    homepage = "https://github.com/SergiusTheBest/plog";
    license = licenses.mit;
    platforms = platforms.linux ++ platforms.darwin;
    maintainers = with maintainers; [ __TODO__ ];
  };
}
```

</details>

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:36]:
```suggestion
```
nlohmann_json is only required by the examples

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (raphaelr) [inline pkgs/development/libraries/libdatachannel/default.nix:46]:
```suggestion
    "-DUSE_NICE=ON"
    "-DUSE_SYSTEM_SRTP=ON"
    "-DNO_EXAMPLES=ON"
```
- USE_SYSTEM_JSON/USE_SYSTEM_USRSCTP only exist on the master branch for now
- I'd skip building the examples for now (since it currently requires a vendored nlohmann_json and we don't install them anyway)

**Review by raphaelr — COMMENTED** [review]
Most feedback is about reducing vendoring

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (erdnaxe) [reply pkgs/development/libraries/libdatachannel/default.nix]: Thanks! This was the fix I was searching for.

**Comment by erdnaxe** [comment]
Using `usrsctp` from NixPkgs (rather than the vendored one) breaks WHIP audio: I get strange audio cracks when streaming to a WHIP server with Opus.

I am going to investigate why before continuing this PR.

Thank you for the plog derivation, I will open another pull request.

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (misuzu) [reply pkgs/development/libraries/libdatachannel/default.nix:43]: > The advantage is that if we update usrsctp in nixpkgs, then libdatachannel will also get that updated version.

There's no new releases for usrsctp, so that's not a good idea.

Comment (erdnaxe) [reply pkgs/development/libraries/libdatachannel/default.nix:58]: Thanks, I added a derivation with libdatachannel usrsctp version.

**Review by Janik-Haag — APPROVED** [review]
LGTM; I will merge this once some one else approves this as well or ~2 weeks go by with out some one complaining.

**Review by raphaelr — APPROVED** [review]
LGTM

File: pkgs/development/libraries/libdatachannel/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/libdatachannel/default.nix:18]: A more canonical name we would use would be ``usrsctp'``

Comment (Janik-Haag) [reply pkgs/development/libraries/libdatachannel/default.nix:18]: we can revert this If you find it more then a slight annoyance.

Comment (erdnaxe) [reply pkgs/development/libraries/libdatachannel/default.nix:18]: I can open another PR if you say this is important.

Comment (SuperSandro2000) [reply pkgs/development/libraries/libdatachannel/default.nix:18]: Just keep it in mind for next time or the next change to this file

**Pull request merged**
